feat!: rework consumer and producer creation

JKRhb opened the pull request:
This PR introduces three subclasses to fulfill the roles of consumer, producer, and a combined consumer/producer ("prosumer"). They inherit from the common `IceFlow` base class, which is turned abstract now in the sense that both its constructor and the methods for pushing and receiving data are now labelled as protected.

I think the general concept should point into the right direction, however, we might need to discuss the details a bit more when we meet the next time. After we resolved this discussion, we can think about a more dynamic way of subscribing and publishing to topics, as well as an API to modify the number of partitions at runtime.

pulsastrix reviewed (approved):
LGTM.

Regarding possible further improvements: 
IMHO, it would be best to separate an instance of "being subscribed to/publisher for a specific topic" from an instance of "being part of the network".

My idea would be to modify the `IceFlow` class in such a way that it allows multiple subscriptions/publishers to be attached to a single instance of IceFlow, and then provide methods to create `Consumer` and `Producer` instances from this "global" instance of `IceFlow`.
This way, publishing and subscribing could be done dynamically, but the central `IceFlow` instance could still maintain some necessary shared state (global input/output queue, SVS instance, NDN Face)

JKRhb commented:
Ah, thank you for your feedback, that sounds good to me :) I am wondering: Should we create a separate "prosumer" class then or should limit ourselves to only introducing separate consumer and producer classes (which would probably be instantiated by `createProducer` or `createConsumer` methods, right?). I feel like having only the two classes would probably decrease the complexity a lot if we go down this route.

pulsastrix commented:
For now, I think that having a `Prosumer` class is fine to make implementing some use cases a bit easier, but later on (when we get around to doing the further improvements) we'd probably only have producers and consumers.

OTOH, one could also argue that it is unnecessary to introduce a class that we already plan on removing, so we could probably go either way.

tanim-ics commented:
> For now, I think that having a `Prosumer` class is fine to make implementing some use cases a bit easier, but later on (when we get around to doing the further improvements) we'd probably only have producers and consumers.
> 
> OTOH, one could also argue that it is unnecessary to introduce a class that we already plan on removing, so we could probably go either way.

I like the "Prosumer" idea most instead of creating 2 different classes. We can avoid letting the application developer from creating 2 class objects as Producer and Consumer. 

@JKRhb - instead of "Prosumer", can we use "streamSession"? As IceFlow is a stream processing library, stream.pushData() and stream.receiveData() sounds a bit more meaningful.

JKRhb commented:
> @JKRhb - instead of "Prosumer", can we use "streamSession"? As IceFlow is a stream processing library, stream.pushData() and stream.receiveData() sounds a bit more meaningful.

Good point, maybe we could also have a naming scheme like `InputStreamSession` (for the producer), `OutputStreamSession` (for the consumer), and `DuplexStreamSession` (for the prosumer, if we want to keep it).

JKRhb commented:
For now, I've started with a rework as suggested by @pulsastrix in https://github.com/hsel-netsys/iceflow/pull/52#pullrequestreview-2006701040. As the rework is still WIP, I've converted the PR to draft for now.

tanim-ics commented:
> > @JKRhb - instead of "Prosumer", can we use "streamSession"? As IceFlow is a stream processing library, stream.pushData() and stream.receiveData() sounds a bit more meaningful.
> 
> Good point, maybe we could also have a naming scheme like `InputStreamSession` (for the producer), `OutputStreamSession` (for the consumer), and `DuplexStreamSession` (for the prosumer if we want to keep it).

yes, that also sounds pretty good actually.

JKRhb commented:
The PR should now be ready for a final review :) Thank you for your patience.

JKRhb commented:
Just noticed that there was one thing missing: The deregistering of producers. Now the PR should actually be ready for review.

JKRhb commented:
I now also added an unsubscribe mechanism for consumers and performed some refactoring that also addresses suggestions by the linter for improved performance.

tanim-ics commented on include/iceflow/producer.hpp:
```diff
@@ -0,0 +1,123 @@
+    return {
+        m_pubTopic,
+        partitionNumber,
+        data,
```
why "," after data? is it syntax or typo?

JKRhb commented:
@pulsastrix @tanim-ics Can you maybe have a final look at this PR and #53? :)

pulsastrix commented on include/iceflow/producer.hpp:
```diff
@@ -0,0 +1,123 @@
+    int partitionIndex = m_partitionCount++ % m_topicPartitions.size();
+    auto partitionNumber = m_topicPartitions[partitionIndex];
+    auto data = m_outputQueue.waitAndPopValue();
+    m_lastPublishTimePoint = std::chrono::steady_clock::now();
```
Identifying the time of the last publish based on when `popQueueValue()` was called might lead to the actual publish interval being shorter than the defined publish interval if the thread is paused after retrieving the queue element, but before actually sending the data.

This _might_ be fine, depending on whether we consider the publish interval a strict minimum or only a recommendation by the user.

If it should be a strict minimum, we should define a callback for the producer that is called after actually publishing the data, which then updates the last publish time.

JKRhb replied:
> If it should be a strict minimum, we should define a callback for the producer that is called after actually publishing the data, which then updates the last publish time.

Yeah, that definitely makes sense, I will update the PR accordingly soon :) I also now remembered another slight change I wanted to make in this part of the code which I will add in the same wake. I will ping you once again as soon as these changes are implemented :)

JKRhb replied:
I hope I now improved the implementation in the latest commit – looking forward to your feedback on Monday :)
